File: typescript/nodejs/src/inspect.ts
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/**
 * The record-by-record walk: what is in a file, where, how long, and whether the file's
 * own checksum covers it.
 *
 * Framing only. A record's content is never read, so this costs the same on a scene
 * carrying a six-megabyte audio payload as on one carrying none, and an opcode nobody here
 * has heard of is stepped over by its own declared length — the whole forward-compatibility
 * story, exercised rather than described.
 *
 * The one thing it does read is the tail: the Footer names the byte range its `summary_crc`
 * covers, and "which records are under the checksum, and did the checksum verify" is not
 * answerable from framing alone. That is one bounded read plus a streamed pass over the
 * summary, never the whole file.
 */

import {
  Crc32,
  Cursor,
  FOOTER_TAIL_BYTES,
  FourdgsError,
  MAGIC,
  Opcode,
  RECORD_HEADER_BYTES,
  bytesEqual,
  checkMagic,
  opcodeName,
  parseFooter,
  type IReadable,
} from "@4dgs/core";

/** One record's framing, and its standing with the file's checksum. */
export interface InspectedRecord {
  readonly opcode: number;
  /** The opcode's registry name, or `Unknown(0xNN)` / `Private(0xNN)`. */
  readonly name: string;
  /** Offset of the record's opcode byte. */
  readonly offset: number;
  /** Content bytes, excluding the nine-byte record header. */
  readonly contentLength: number;
  /** Total bytes the record occupies, header included. */
  readonly totalLength: number;
  /** True when this record lies inside the range the Footer's `summary_crc` covers. */
  readonly covered: boolean;
}

/** The summary checksum, as the Footer declares it and as the bytes actually compute. */
export interface SummaryCrc {
  /** First byte covered — the Footer's `summary_start`. */
  readonly start: number;
  /** One past the last byte covered: the Footer record's own first byte. */
  readonly end: number;
  readonly declared: number;
  readonly actual: number;
  readonly ok: boolean;
}

export interface InspectReport {
  readonly size: number;
  readonly records: readonly InspectedRecord[];
  /** True when the last eight bytes are the magic, as a complete file's are. */
  readonly trailingMagic: boolean;
  /**
   * Why the walk stopped early, or `null` when it reached the end.
   *
   * A cut file is reported, not thrown: everything framed before the cut is what the
   * decoder would also recover, and hiding it behind an exception would answer "this file
   * is broken" to a question that was "how far in does it break".
   */
  readonly stopped: string | null;
  /** `null` when the file's summary checksum could not be reached or was never written. */
  readonly summaryCrc: SummaryCrc | null;
  /**
   * Why there is no checksum to report, when there is none — and `null` when there is one.
   *
   * A file that declares no checksum and a file whose tail was cut off both have nothing to
   * verify, and saying so the same way would be the tool misreading the second as the first.
   */
  readonly summaryCrcAbsent: string | null;
}

/** Bytes per read while checksumming the summary. Bounds what is in flight, not what fits. */
const CRC_BLOCK_BYTES = 64 * 1024;

/**
 * Walk `source` record by record.
 *
 * Throws only what the magic check throws — a file that is not ours, or is from a major
 * version this reader does not implement, has no records to walk. Everything after that is
 * reported rather than raised.
 */
export async function inspectFile(source: IReadable): Promise<InspectReport> {
  const size = Number(await source.size());
  checkMagic(await source.read(0n, BigInt(Math.min(MAGIC.length, size))));

  const summary = await readSummaryCrc(source, size);
  const summaryCrc = typeof summary === "string" ? null : summary;
  const summaryCrcAbsent = typeof summary === "string" ? summary : null;
  const records: InspectedRecord[] = [];
  let at = MAGIC.length;
  let trailingMagic = false;
  let stopped: string | null = null;

  for (;;) {
    const remaining = size - at;
    if (remaining <= 0) break;
    // A complete file ends with the magic, which is shorter than a record header and so is
    // never framed as one.
    if (remaining <= MAGIC.length) {
      trailingMagic = bytesEqual(await source.read(BigInt(at), BigInt(remaining)), MAGIC);
      // Anything else in that space is a file that was cut inside its own closing magic:
      // the records all framed, so nothing above reported a stop, and the tail is short by
      // a byte or seven. Saying so here is what keeps `inspect` from exiting 0 on a file
      // the validator refuses one line later for the same reason — "file does not end with
      // the magic" is a finding there, and it was a note here.
      if (!trailingMagic) {
        stopped = `${remaining} bytes at ${at} are not the closing magic; the file is truncated`;
      }
      break;
    }
    if (remaining < RECORD_HEADER_BYTES) {
      stopped = `${remaining} bytes at ${at} are not a record header`;
      break;
    }
    const framing = new Cursor(await source.read(BigInt(at), BigInt(RECORD_HEADER_BYTES)));
    const opcode = framing.u8();
    let contentLength: number;
    try {
      contentLength = framing.u64();
    } catch (error) {
      // A length past 2^53 is a number this runtime cannot hold exactly, and an offset
      // rounded to the nearest even byte reads the wrong record rather than failing.
      stopped = error instanceof FourdgsError ? error.message : String(error);
      break;
    }
    const totalLength = contentLength + RECORD_HEADER_BYTES;
    const end = at + totalLength;
    if (!Number.isSafeInteger(end) || end > size) {
      stopped =
        `the ${opcodeName(opcode)} record at ${at} declares ${contentLength} bytes, ` +
        `past the end of a ${size}-byte file`;
      records.push(row(opcode, at, contentLength, totalLength, summaryCrc));
      break;
    }
    records.push(row(opcode, at, contentLength, totalLength, summaryCrc));
    at = end;
  }

  return { size, records, trailingMagic, stopped, summaryCrc, summaryCrcAbsent };
}

function row(
  opcode: number,
  offset: number,
  contentLength: number,
  totalLength: number,
  summary: SummaryCrc | null,
): InspectedRecord {
  return {
    opcode,
    name: opcodeName(opcode),
    offset,
    contentLength,
    totalLength,
    covered: summary !== null && offset >= summary.start && offset < summary.end,
  };
}

/**
 * The Footer's declared summary checksum, and what the covered bytes actually compute to.
 *
 * A string instead for every file that does not answer the question: one too short to hold
 * a footer, one whose tail is not a Footer record, one that declares no checksum (legal —
 * the field is zero when a writer did not compute one), and one whose declared range is
 * impossible. A checksum nobody can check is not a failed checksum, and each of those is a
 * different reason worth saying out loud.
 */
async function readSummaryCrc(source: IReadable, size: number): Promise<SummaryCrc | string> {
  if (size < MAGIC.length + FOOTER_TAIL_BYTES) {
    return "the file is too short to hold a Footer record";
  }
  const end = size - FOOTER_TAIL_BYTES;
  const tail = await source.read(BigInt(end), BigInt(FOOTER_TAIL_BYTES));
  if (tail[0] !== Opcode.Footer) return "there is no Footer record at the tail";
  let footer;
  try {
    footer = parseFooter(tail.subarray(RECORD_HEADER_BYTES));
  } catch {
    return "the Footer record does not parse";
  }
  if (footer.summaryCrc === 0 || footer.summaryStart === 0) return "the file declares none";
  if (footer.summaryStart > end) {
    return `the Footer's summary starts at ${footer.summaryStart}, after the summary ends at ${end}`;
  }

  const crc = new Crc32();
  for (let at = footer.summaryStart; at < end; at += CRC_BLOCK_BYTES) {
    crc.update(await source.read(BigInt(at), BigInt(Math.min(CRC_BLOCK_BYTES, end - at))));
  }
  const actual = crc.digest();
  return {
    start: footer.summaryStart,
    end,
    declared: footer.summaryCrc,
    actual,
    ok: actual === footer.summaryCrc,
  };
}

File: dart/fourdgs/lib/src/inspect.dart
// Copyright 2026 Avala AI
// SPDX-License-Identifier: Apache-2.0

/// Walk the records: opcode by name, byte offset, length, CRC status.
///
/// Framing only, plus the summary region the Footer names. A record's content is
/// never read, so this costs the same on a file with an embedded audio payload as
/// on one without — and an opcode nobody here has heard of is stepped over by its
/// own declared length, which is the whole forward-compatibility story, exercised
/// rather than described.
///
/// A file that was cut is walked as far as it goes and then says so. That is what
/// the decoder does with one — records are length-prefixed, so everything
/// complete before the cut is intact and the streamed reader keeps it — and a
/// tool whose job is to say where a file stops being a 4dgs file should not
/// answer that question by printing one line and throwing the rest away.
library;

import 'exceptions.dart';
import 'opcode.dart';
import 'readable.dart';
import 'records.dart';
import 'refusal.dart';
import 'serialization.dart';

/// The only checksum the format defines: `summary_crc` over the bytes from
/// `summary_start` to where the Footer begins.
///
/// So a record's "CRC status" is a fact about the region it sits in rather than a
/// field of its own — covered and agreeing, covered and not, or covered by
/// nothing at all — and saying so per record is what tells a reader whether the
/// checksum has anything to say about the record they are looking at.
class FourdgsSummaryCoverage {
  const FourdgsSummaryCoverage({
    required this.start,
    required this.end,
    required this.ok,
  });

  final int start;

  /// One past the last covered byte: where the Footer record's opcode sits.
  final int end;

  final bool ok;

  /// The cell for a record, in the vocabulary `validate` already uses.
  String cell(int at, int total) {
    if (at < start || at + total > end) return '-';
    return ok ? 'ok' : 'MISMATCH';
  }
}

/// Every record a file frames, and what the summary checksum covers.
class FourdgsInspection {
  const FourdgsInspection({
    required this.walk,
    required this.coverage,
    this.coverageError,
  });

  final FourdgsWalk walk;

  /// `null` when the file declares no summary checksum, which is a property of
  /// the file rather than a failure: writing one is an encoder option, and a file
  /// written without it has nothing here to verify.
  final FourdgsSummaryCoverage? coverage;

  /// Why there is no [coverage] when the reason was a fault rather than a
  /// choice: a Footer that frames cleanly and will not parse.
  ///
  /// Reported instead of thrown. The framing walk is the answer this command
  /// exists to give, and a record that is itself malformed is exactly the case
  /// its holder ran `inspect` for — throwing here would print no table at all
  /// for the one file that most needs one.
  final String? coverageError;
}

/// Frames every record, then checks the summary region the Footer declares.
Future<FourdgsInspection> inspectFourdgs(FourdgsReadable source) async {
  final FourdgsWalk walk = await walkFourdgsFraming(source);
  try {
    return FourdgsInspection(
      walk: walk,
      coverage: await _coverage(source, walk),
    );
  } on FourdgsException catch (error) {
    return FourdgsInspection(
      walk: walk,
      coverage: null,
      coverageError: error.message,
    );
  }
}

Future<FourdgsSummaryCoverage?> _coverage(
  FourdgsReadable source,
  FourdgsWalk walk,
) async {
  final FourdgsFrame? frame = walk.first(opFooter);
  if (frame == null || frame.offset + frame.total > walk.size) return null;
  final FourdgsFooter footer = FourdgsFooter.parse(
    await source.read(frame.offset + recordHeaderBytes, frame.length),
  );
  if (footer.summaryCrc == 0 ||
      footer.summaryStart == 0 ||
      footer.summaryStart > frame.offset) {
    return null;
  }
  // The summary ends where the Footer begins — taken from the walk rather than
  // computed from a Footer's expected size, so a Footer that a later revision
  // extends does not move the region out from under the check.
  //
  // Checked a block at a time. The region is whatever the Footer says it is, and
  // a Footer is an untrusted field on a file this command is pointed at
  // *because* it is suspect: `summary_start` just past the magic makes the
  // region almost the whole file, and reading it in one allocation is a crafted
  // Footer away from exhausting memory to check a checksum (AGENTS.md §1).
  final int actual = await fourdgsCrc32Range(
    source,
    footer.summaryStart,
    frame.offset - footer.summaryStart,
  );
  return FourdgsSummaryCoverage(
    start: footer.summaryStart,
    end: frame.offset,
    ok: actual == footer.summaryCrc,
  );
}

/// The record table, in the column widths the Rust and TypeScript tools print.
///
/// Same columns in the same order, because the point of six tools that read one
/// format is that their output can be diffed rather than interpreted.
String formatFourdgsInspection(FourdgsInspection inspection) {
  final FourdgsWalk walk = inspection.walk;
  final FourdgsSummaryCoverage? coverage = inspection.coverage;
  final StringBuffer out = StringBuffer();
  void row(String a, String b, String c, String d, String e) => out.writeln(
    '${a.padLeft(12)}  ${b.padRight(18)} ${c.padLeft(14)}  ${d.padLeft(14)}  $e',
  );

  row('offset', 'record', 'content', 'total', 'crc');
  row('0', '(magic)', '', '8', '-');
  for (final FourdgsFrame frame in walk.records) {
    row(
      fourdgsCommas(frame.offset),
      opcodeName(frame.opcode),
      fourdgsCommas(frame.length),
      fourdgsCommas(frame.total),
      coverage?.cell(frame.offset, frame.total) ?? '-',
    );
  }
  if (walk.trailingMagic) {
    row(fourdgsCommas(walk.size - 8), '(magic)', '', '8', '-');
  }

  out.writeln();
  out.writeln(
    '${walk.records.length} records, ${fourdgsCommas(walk.size)} bytes',
  );
  final FourdgsCut? cut = walk.cut;
  if (cut != null) {
    out.writeln('truncated at byte ${fourdgsCommas(cut.at)}: ${cut.reason}');
    final String last =
        cut.insideARecord
            ? '; the last row is the record the file was cut inside'
            : '';
    out.writeln(
      'the ${walk.intact} complete records above are the intact prefix, which '
      'is what a streamed reader keeps$last',
    );
  } else if (!walk.trailingMagic) {
    out.writeln('note: the file does not end with the magic');
  }
  final String? coverageError = inspection.coverageError;
  if (coverageError != null) {
    out.writeln(
      'crc: the Footer frames cleanly but does not parse ($coverageError), so '
      'nothing here could be checked',
    );
  } else if (coverage == null) {
    out.writeln(
      'crc: this file declares no summary checksum, so nothing here is covered',
    );
  } else {
    out.writeln(
      "crc: the Footer's summary checksum covers bytes "
      '${fourdgsCommas(coverage.start)}..${fourdgsCommas(coverage.end)}; '
      '`-` is a record it does not cover',
    );
  }
  return out.toString();
}

/// The same walk, for a caller that is a program rather than a person.
///
/// Written by hand rather than through `dart:convert` so that the field names and
/// their order are the Rust tool's, key for key: a script that reads one tool's
/// `--json` reads the other's.
String formatFourdgsInspectionJson(FourdgsInspection inspection) {
  final FourdgsWalk walk = inspection.walk;
  final FourdgsSummaryCoverage? coverage = inspection.coverage;
  final FourdgsCut? cut = walk.cut;
  final StringBuffer out = StringBuffer();
  out.writeln('{');
  out.writeln('  "size": ${walk.size},');
  out.writeln('  "trailing_magic": ${walk.trailingMagic},');
  if (cut == null) {
    out.writeln('  "stopped": null,');
    out.writeln('  "truncated_at": null,');
  } else {
    out.writeln('  "stopped": ${_jsonString(cut.reason)},');
    out.writeln('  "truncated_at": ${cut.at},');
  }
  final String? coverageError = inspection.coverageError;
  if (coverageError != null) {
    out.writeln('  "summary_crc": null,');
    out.writeln('  "summary_crc_error": ${_jsonString(coverageError)},');
  } else if (coverage == null) {
    out.writeln('  "summary_crc": null,');
  } else {
    out.writeln(
      '  "summary_crc": {"start": ${coverage.start}, '
      '"end": ${coverage.end}, "ok": ${coverage.ok}},',
    );
  }
  out.writeln('  "records": [');
  for (int i = 0; i < walk.records.length; i++) {
    final FourdgsFrame frame = walk.records[i];
    final String cell = coverage?.cell(frame.offset, frame.total) ?? '-';
    final String crc = cell == '-' ? 'null' : _jsonString(cell.toLowerCase());
    final String comma = i + 1 == walk.records.length ? '' : ',';
    out.writeln(
      '    {"offset": ${frame.offset}, "opcode": ${frame.opcode}, '
      '"name": ${_jsonString(opcodeName(frame.opcode))}, '
      '"content_length": ${frame.length}, "total_length": ${frame.total}, '
      '"crc": $crc}$comma',
    );
  }
  out.writeln('  ]');
  out.writeln('}');
  return out.toString();
}

/// A string as a JSON string. The only escapes a record's fields can need are
/// these: opcode names are ASCII and a cut's reason is this file's own prose.
String _jsonString(String s) {
  final StringBuffer out = StringBuffer('"');
  for (final int unit in s.runes) {
    if (unit == 0x22) {
      out.write(r'\"');
    } else if (unit == 0x5C) {
      out.write(r'\\');
    } else if (unit == 0x0A) {
      out.write(r'\n');
    } else if (unit < 0x20) {
      out.write('\\u${unit.toRadixString(16).padLeft(4, '0')}');
    } else {
      out.writeCharCode(unit);
    }
  }
  out.write('"');
  return out.toString();
}
